cli/src/commands/upload.ts

```diff
@@ -0,0 +1,183 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+
+import { BaseCommand } from '../cli/base-command';
+import { CrawledAsset } from '../cores/models/crawled-asset';
+import { CrawlService, UploadService } from '../services';
+import * as si from 'systeminformation';
+import FormData from 'form-data';
+import { UploadOptionsDto } from '../cores/dto/upload-options-dto';
+import { CrawlOptionsDto } from '../cores/dto/crawl-options-dto';
+
+export default class Upload extends BaseCommand {
+  private crawlService = new CrawlService();
+  private uploadService!: UploadService;
+  deviceId!: string;
+  uploadLength!: number;
+  dryRun = false;
+
+  public async run(paths: string[], options: UploadOptionsDto): Promise<void> {
+    await this.connect();
+
+    const uuid = await si.uuid();
+    this.deviceId = uuid.os || 'CLI';
+    this.uploadService = new UploadService(this.immichApi.apiConfiguration);
+
+    this.dryRun = options.dryRun;
+
+    const crawlOptions = new CrawlOptionsDto();
+    crawlOptions.pathsToCrawl = paths;
+    crawlOptions.recursive = options.recursive;
+    crawlOptions.excludePatterns = options.excludePatterns;
+
+    const crawledFiles: string[] = await this.crawlService.crawl(crawlOptions);
+
+    if (crawledFiles.length === 0) {
+      console.log('No assets found, exiting');
+      return;
+    }
+
+    const assetsToUpload = crawledFiles.map((path) => new CrawledAsset(path));
+
+    const cliProgress = require('cli-progress');
+    const byteSize = require('byte-size');
+
+    const uploadProgress = new cliProgress.SingleBar(
+      {
+        format: '{bar} | {percentage}% | ETA: {eta_formatted} | {value_formatted}/{total_formatted}: {filename}',
+      },
+      cliProgress.Presets.shades_classic,
+    );
+
+    let totalSize = 0;
+    let sizeSoFar = 0;
+
+    let totalSizeUploaded = 0;
+    let uploadCounter = 0;
+
+    for (const asset of assetsToUpload) {
+      // Compute total size first
+      await asset.process();
+      totalSize += asset.fileSize;
+    }
+
+    uploadProgress.start(totalSize);
+    uploadProgress.update({ value_formatted: 0, total_formatted: byteSize(totalSize) });
+
+    for (const asset of assetsToUpload) {
+      uploadProgress.update({
+        filename: asset.path,
+      });
+
+      try {
+        if (options.import) {
+          await this.importAsset(asset);
+        } else {
+          await this.uploadAsset(asset, options.skipHash);
+        }
+      } catch (error) {
+        uploadProgress.stop();
+        throw error;
+      }
+
+      sizeSoFar += asset.fileSize;
+      if (!asset.skipped) {
+        totalSizeUploaded += asset.fileSize;
+        uploadCounter++;
+      }
+
+      uploadProgress.update(sizeSoFar, { value_formatted: byteSize(sizeSoFar) });
+    }
+
+    uploadProgress.stop();
+
+    if (options.import) {
```

Comment:
I think we can simple this very nested if/else set of conditions. At least for the dry run stuff it looks like the messages are basically identical. Maybe we can get a message prefix of `Would have` vs `Successfully` and prepend that the the messages instead of using an if/else in some places.

Reply:
Yes, this turned out a bit messy in the end